### PR TITLE
Raising metric events to track individual function executions, refactoring MetricsEventManager and adding unit test cases

### PR DESCRIPTION
**Pull request by akurmi**

### CODE REVIEW

**Comment by akurmi** [comment]
@ShekharGupta1988

File: src/WebJobs.Script.WebHost/Diagnostics/WebHostMetricsLogger.cs [hunk 1]
```diff
@@ -8,13 +8,25 @@ namespace WebJobs.Script.WebHost.Diagnostics
 {
     public class WebHostMetricsLogger : IMetricsLogger
     {
+        private MetricsEventManager metricsEventManager;
```
Comment (mathewc) [inline src/WebJobs.Script.WebHost/Diagnostics/WebHostMetricsLogger.cs]: nit: in this repo, we prefix all private members with a _

File: test/WebJobs.Script.Tests/MetricsEventManagerTests.cs [hunk 2]
```diff
@@ -0,0 +1,357 @@
...
+using Microsoft.Azure.WebJobs.Script.Description;
+using Microsoft.Azure.WebJobs.Script.Diagnostics;
+using Moq;
+using WebJobs.Script.WebHost.Diagnostics;
+using WebJobs.Script.WebHost.Models;
+using Xunit;
+
+namespace Microsoft.Azure.WebJobs.Script.Tests
+{
+    public class MetricsEventManagerTests
```
Comment (mathewc) [inline test/WebJobs.Script.Tests/MetricsEventManagerTests.cs]: Awesome to have these tests in place :)

File: test/WebJobs.Script.Tests/MetricsEventManagerTests.cs [hunk 3]
```diff
@@ -0,0 +1,357 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
```
Comment (mathewc) [inline test/WebJobs.Script.Tests/MetricsEventManagerTests.cs]: nit: remove extraneous space

File: test/WebJobs.Script.Tests/MetricsEventManagerTests.cs [hunk 4]
```diff
@@ -0,0 +1,357 @@
...
+                errorMessage = string.Format("No of finished events does not match with number of function executions: Expected:{0} Actual:{1}", noOfFuncExecutions, hashes.Count);
+                return false;
+            }
+
+            return true;
+        }
+
+        private static WebHostMetricsLogger CreateWebHostMetricsLoggerInstance(List<FunctionExecutionEventArguments> functionExecutionEventArgumentsList)
+        {
+            var metricsEventGenerator = new Mock<IMetricsEventGenerator>();
```
Comment (mathewc) [inline test/WebJobs.Script.Tests/MetricsEventManagerTests.cs]: You don't need Moq for this - recommend simply implementing your interface to collect the events

File: test/WebJobs.Script.Tests/MetricsEventManagerTests.cs [hunk 5]
```diff
@@ -0,0 +1,357 @@
...
+                .Callback(
+                (string executionId, string siteName, int concurrency, string functionName, string invocationId, string executionStage, long executionTimeSpan, bool success) =>
+                {
+                    functionExecutionEventArgumentsList.Add(new FunctionExecutionEventArguments(executionId, siteName, concurrency, functionName, invocationId, executionStage, executionTimeSpan, success));
+                });
+
+            return new WebHostMetricsLogger(metricsEventGenerator.Object);
+        }
+
+        private async Task LongDummyFunction(WebHostMetricsLogger metricsLogger)
```
Comment (mathewc) [inline test/WebJobs.Script.Tests/MetricsEventManagerTests.cs]: nit: can you rename Dummy to Test? (LongTestFunction/ShortTestFunction)

File: src/WebJobs.Script.WebHost/Diagnostics/IMetricsEventGenerator.cs [hunk 6]
```diff
@@ -0,0 +1,22 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
```
Comment (mathewc) [inline src/WebJobs.Script.WebHost/Diagnostics/IMetricsEventGenerator.cs]: nit: run "Sort and order usings" on all these files to remove unused usings

**Comment by mathewc** [comment]
Merged. Please address the test timing issue and above comments in a follow up PR, thanks :)

**Pull request merged**